network/src/topology.rs

```diff
@@ -0,0 +1,492 @@
+///
+/// This is defined as the connection count of the node.
+fn degree_centrality(index: &BTreeMap<SocketAddr, usize>, degree_matrix: &DMatrix<f64>) -> BTreeMap<SocketAddr, u16> {
+    let diag = degree_matrix.diagonal();
+    index
+        .keys()
+        .zip(diag.iter())
+        .map(|(addr, dc)| (*addr, *dc as u16))
+        .collect()
+}
+
+/// Returns the eigenvalue centrality of each node in the network.
+fn eigenvector_centrality(
+    index: &BTreeMap<SocketAddr, usize>,
+    adjacency_matrix: DMatrix<f64>,
+) -> BTreeMap<SocketAddr, f64> {
+    // Compute the eigenvectors and corresponding eigenvalues and sort in descending order.
+    let ascending = false;
+    let eigenvalue_vector_pairs = sorted_eigenvalue_vector_pairs(adjacency_matrix, ascending);
+    let (_highest_eigenvalue, highest_eigenvector) = &eigenvalue_vector_pairs[0];
+
+    // The eigenvector is a relative score of node importance (normalised by the norm), to obtain an absolute score for each
+    // node, we normalise so that the sum of the components are equal to 1.
+    let sum = highest_eigenvector.sum() / index.len() as f64;
+    let normalised = highest_eigenvector.unscale(sum);
+
+    // Map addresses to their eigenvalue centrality.
+    index
+        .keys()
+        .zip(normalised.column(0).iter())
+        .map(|(addr, ec)| (*addr, *ec))
+        .collect()
+}
+
+/// Returns the Fiedler values for each node in the network.
+fn fiedler(index: &BTreeMap<SocketAddr, usize>, laplacian_matrix: DMatrix<f64>) -> (f64, BTreeMap<SocketAddr, f64>) {
+    // Compute the eigenvectors and corresponding eigenvalues and sort in ascending order.
+    let ascending = true;
+    let pairs = sorted_eigenvalue_vector_pairs(laplacian_matrix, ascending);
+
+    // Second-smallest eigenvalue is the Fiedler value (algebraic connectivity), the associated
+    // eigenvector is the Fiedler vector.
+    let (algebraic_connectivity, fiedler_vector) = &pairs[1];
+
+    // Map addresses to their Fiedler values.
+    let fiedler_values_indexed = index
+        .keys()
+        .zip(fiedler_vector.column(0).iter())
+        .map(|(addr, fiedler_value)| (*addr, *fiedler_value))
+        .collect();
+
+    (*algebraic_connectivity, fiedler_values_indexed)
+}
+
+/// Computes the eigenvalues and corresponding eigenvalues from the supplied symmetric matrix.
+fn sorted_eigenvalue_vector_pairs(matrix: DMatrix<f64>, ascending: bool) -> Vec<(f64, DVector<f64>)> {
+    // Compute eigenvalues and eigenvectors.
+    let eigen = SymmetricEigen::new(matrix);
+
+    // Map eigenvalues to their eigenvectors.
+    let mut pairs: Vec<(f64, DVector<f64>)> = eigen
+        .eigenvalues
+        .iter()
+        .zip(eigen.eigenvectors.column_iter())
+        .map(|(value, vector)| (*value, vector.clone_owned()))
+        .collect();
+
+    // Sort eigenvalue-vector pairs in descending order.
+    pairs.sort_unstable_by(|(a, _), (b, _)| {
+        if ascending {
+            a.partial_cmp(b).unwrap()
```

Comment:
are we certain we can't encounter any `f64::NAN` values that would panic here?

Reply:
I'd be surprised but I'll check `nalgebra`.

Reply:
The algorithm should be safe so long as the matrix is symmetric (guaranteed by the type). `nalgebra` `unwrap`s the `SymmetricEigen::new`, there's nothing else that could result in an invalid value at that point. I'll thoroughly test this computation before exposing it via rpc (likely a follow up) just to be sure but so far haven't encountered any issues.